src/caffe/layers/memory_data_layer.cpp

```diff
@@ -34,9 +35,12 @@ void MemoryDataLayer<Dtype>::AddDatumVector(const vector<Datum>& datum_vector) {
       << " by the upper layers";
   size_t num = datum_vector.size();
   CHECK_GT(num, 0) << "There is no datum to add";
-  CHECK_LE(num, batch_size_) <<
-      "The number of added datum must be no greater than the batch size";
-
+  CHECK_LE(num % batch_size_, 0) <<
+      "The number of added datum must be multiple of the batch size";
+  if (num > batch_size_) {
```

Comment:
`added_data_` and `added_label_` should be reshaped to `num` always, not only when bigger than `batch_size_`.
Since Blob.Reshape is lazy it will not do anything at least it is needed (i.e. increasing the size)